pennylane/plugins/expt_tensornet.py

```diff
@@ -0,0 +1,197 @@
+# Copyright 2018 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+Experimental simulator plugin based on tensor network contractions
+"""
+import functools
+
+import numpy as np
+import tensornetwork as tn
+
+from pennylane.utils import _flatten
+from pennylane import Device
+from pennylane.plugins.default_qubit import I, X, Y, Z, H, CNOT, SWAP, CZ, S, T, CSWAP, \
+    Rphi, Rotx, Roty, Rotz, Rot3, CRot3, CRotx, CRoty, CRotz, CRot3, hermitian, identity, \
+    unitary
+
+# tolerance for numerical errors
+tolerance = 1e-10
+
+#========================================================
+#  device
+#========================================================
+
+
+class TensorNetwork(Device):
+    """Experimental Tensor Network simulator device for PennyLane.
+
+    Args:
+        wires (int): the number of modes to initialize the device in
+    """
+    name = 'PennyLane TensorNetwork simulator plugin'
+    short_name = 'expt.tensornet'
+    pennylane_requires = '0.7'
+    version = '0.7.0'
+    author = 'Xanadu Inc.'
+    _capabilities = {"model": "qubit", "tensor_observables": True}
+
+    _operation_map = {
+        'QubitStateVector': None,
+        'BasisState': None,
+        'QubitUnitary': unitary,
+        'PauliX': X,
+        'PauliY': Y,
+        'PauliZ': Z,
+        'Hadamard': H,
+        'S': S,
+        'T': T,
+        'CNOT': CNOT,
+        'SWAP': SWAP,
+        'CSWAP':CSWAP,
+        'CZ': CZ,
+        'PhaseShift': Rphi,
+        'RX': Rotx,
+        'RY': Roty,
+        'RZ': Rotz,
+        'Rot': Rot3,
+        'CRX': CRotx,
+        'CRY': CRoty,
+        'CRZ': CRotz,
+        'CRot': CRot3
+    }
+
+    _observable_map = {
+        'PauliX': X,
+        'PauliY': Y,
+        'PauliZ': Z,
+        'Hadamard': H,
+        'Hermitian': hermitian,
+        'Identity': identity
+    }
+
+    def __init__(self, wires):
+        super().__init__(wires, shots=1)
+        self.eng = None
+        self.analytic = True
+        self._nodes = []
+        self._edges = []
+        self._zero_state = np.zeros([2] * wires)
+        self._zero_state[tuple([0] * wires)] = 1.0
+        # TODO: since this state is separable, can be more intelligent about not making a dense matrix
+        self._state = self._add_node(self._zero_state, wires=tuple(w for w in range(wires)), name="AllZeroState")
+        self._free_edges = [e for e in self._state.edges]  # we need this list to be distinct from self._state.edges
```

Comment:
Might be clearer if this was an explicit `copy()`, rather than a list comprehension. Or maybe `self._free_edges = self._state.edges[:]`?

Reply:
This was a really annoying bug. Initially I set `self._free_edges = self._state_edges`. When I changed something in `self._free_edges` (needed for bookkeeping of tensornetwork), it would screw up the state node

Reply:
But yes, I think `copy` should be sufficient